### ISSUE TITLE
Get regex/string matching to work on Tamil and Hindi.

### ISSUE THREAD

**Issue by tarunima** [tattle-made/Uli#310]
From presentation on Thursday we know that the standard regex library was breaking on certain unicode characters. Thus, approximate slur replacement may not work on Tamil and Hindi. This ticket is to resolve those issues and get approximate matching to work on these languages.

**Comment by duggalsu** [comment]
The current regex slur replacement feature is implemented using JavaScript `RegExp` in `slur-replace.js`. 

The code in `Uli/slur-replacement` and the one used for the presentation mentioned above is implemented using the Python `re` regex engine. 

Both regex engines will need to be tested separately for any issues.

The current JavaScript regex implementation fails in the following cases. These are not Unicode specific issues - 
1. If there is a partial slur match and the partial word exists as a slur earlier in the list provided, the replacement fails
2. Phrases are not getting replaced due to spaces, and if the first word exists as a slur earlier in the list provided
3. If 2 or more slurs exist earlier in the slur list and the current word is written without spaces, there are multiple independent matches and the entire word does not get replaced. 

All Tamil, Hindi and English words are getting replaced without Unicode issues in initial testing.

Steps to reproduce in JavaScript browser console. Code is taken from `slur-replace.js` . We are basically going to iterate over our slur list and see if each slur gets replaced by the RegExp pattern built using the slur list.
```
const PUNCTUATIONS = [' ', '.', ',', ';', '#', '"'];

function generateMask(word, offset, string) {
    const preceding_word = string[offset - 1];
    const next_word = string[offset + word.length];

    if (
        (preceding_word == undefined ||
            PUNCTUATIONS.includes(preceding_word)) &&
        (next_word == undefined || PUNCTUATIONS.includes(next_word))
    ) {
        return '▓'.repeat(word.length);
    } else {
        return word;
    }
}

var slurList = <INSERT slur list>  # this is not really a list but a '|' separated input for RegExp

var expression = new RegExp(`(?:${slurList})`, 'gi');

# New code for testing
var slurList2 = <INSERT slur list>

# create list for iteration to test slur replacement
slurList2 = slurList2.split('|')

for (var i = 0; i < slurList2.length; i++) {
    slurList2[i] = slurList2[i].replace(expression, generateMask);

```

**Comment by duggalsu** [comment]
The issue with regex matching failures was due to longer slur phrases/words appearing later in the list. This resulted in partial matching of smaller slurs. I tried some ideal and non-ideal fixes. PR #313  is an attempt to make near-exact slur matching work in real-time.

The current approach works like this - 
1. Sort the slur list in descending order of phrase length
2. Apply regex replacement
This is currently missing replacing 3 slurs when we try to match our own slur list.
Without sorting the slur list, we are currently missing ~40 slurs.

Other approaches that did not work - 
1. Replacing regex with `replaceAll` was missing multiple occurrences of slurs if surrounded by punctuation or extra unicode characters. This was also slower as it was running in a for-loop. We cannot use web workers for long running slur replacement as it will break core real-time slur replacment functionality.
2. The ideal scenario that I attempted was like this -
i. Split the slur list into an array
ii. Count the slur list phrase length
iii. Create multiple slur lists based on phrase length. We essentially now have a list of lists in descending order of phrase length.
iv. Iterate through this list of lists and `replaceAll` the slurs.

There is serious computational bottleneck with this approach
1. The `replaceAll` slows down real-time slur replacement and also misses matching words surrounded by punctuation and other characters.
2. Running regex instead of replaceAll is still very slow since we are iterating in a loop of lists - the page visibily never gets updated.
